### PR TITLE
Move ellipsis to rlang

### PR DESCRIPTION
**Pull request by lionel-**
Differences compared to the ellipsis package:

- Use r-lib bullets formatting in error messages
- Added `check_dots_empty0()` as a more efficient variant of `check_dots_empty()` for low level functions. Marked as internal since it's mostly meant for rlang and vctrs.

```r
f <- function(x, ..., y) {
  check_dots_empty()
  NULL
}
g <- function(x, ..., y) {
  if (!missing(...)) {
    check_dots_empty()
  }
  NULL
}
h <- function(x, ..., y) {
  check_dots_empty0(...)
  NULL
}

bench::mark(f(), g(), h())[1:8]
#> # A tibble: 3 x 8
#>   expression      min   median `itr/sec` mem_alloc `gc/sec` n_itr
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl> <int>
#> 1 f()           1.5µs   2.58µs   308246.        0B     30.8  9999
#> 2 g()           208ns    333ns  3054071.        0B      0   10000
#> 3 h()           459ns    542ns  1837891.        0B      0   10000
```

### CODE REVIEW

File: R/dots-ellipsis.R [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+#' [check_dots_empty()] with a slightly different interface. Instead
+#' of inspecting the current environment for dots, it directly takes
+#' `...`.  dots. It is only meant for very low level functions where a
+#' couple microseconds make a difference.
+#'
+#' @param ... Dots which should be empty.
+#' @keywords internal
+#' @export
+check_dots_empty0 <- function(...) {
+  nargs()
```
Comment (DavisVaughan) [inline R/dots-ellipsis.R:158]: Is this supposed to error?

File: R/dots-ellipsis.R [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+    note = "These dots only exist to allow future extensions and should be empty.",
+    class = "rlib_error_dots_nonempty"
+  )
+}
+#' Check that dots are empty (low level variant)
+#'
+#' `check_dots_empty0()` is a more efficient version of
+#' [check_dots_empty()] with a slightly different interface. Instead
+#' of inspecting the current environment for dots, it directly takes
+#' `...`.  dots. It is only meant for very low level functions where a
```
Comment (DavisVaughan) [inline R/dots-ellipsis.R:151]:
```suggestion
#' `...`. It is only meant for very low level functions where a
```

File: R/dots-ellipsis.R (hunk 1 above)
Comment (lionel-) [reply R/dots-ellipsis.R:158]: ooops

File: R/dots-ellipsis.R [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+
+exit_handler <- function(action) {
+  expr(
+    on.exit((!!check_dots)(environment(), !!action), add = TRUE)
+  )
+}
+
+#' Check that all dots are unnamed
+#'
+#' In functions like `paste()`, named arguments in ... are often a
```
Comment (DavisVaughan) [inline R/dots-ellipsis.R:75]:
```suggestion
#' In functions like `paste()`, named arguments in `...` are often a
```

File: tests/testthat/_snaps/dots-ellipsis.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+# error if dots named
+
+    Code
+      (expect_error(f(1, 2, 3, xy = 4, x = 5), class = "rlib_error_dots_named"))
+    Output
+      <error/rlib_error_dots_named>
+      2 arguments in `...` had unexpected names.
+      x We detected these problematic arguments:
+      * `xy`
+      * `x`
```
Comment (DavisVaughan) [inline tests/testthat/_snaps/dots-ellipsis.md:15]: Have we ever discussed indenting these sub bullets? I think that might look nice

Comment (lionel-) [reply tests/testthat/_snaps/dots-ellipsis.md:15]: Ah there is an ongoing Slack discussion with @gaborcsardi and @jennybc about this on Slack. We might add and `indent` argument to `format_error_bullets()` and cli might gain special interpolation syntax for this.

**Review by hadley — APPROVED** [review]

**Pull request merged**